processing/src/main/java/org/apache/druid/segment/ColumnarValueUnnestCursor.java

```diff
@@ -0,0 +1,270 @@
+          public long getLong()
+          {
+            Object value = getObject();
+            if (value == null) {
+              return 0;
+            }
+            return Long.valueOf((String) value);
+          }
+
+          @Override
+          public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+          {
+            baseColumSelectorFactory.makeColumnValueSelector(columnName).inspectRuntimeShape(inspector);
+          }
+
+          @Override
+          public boolean isNull()
+          {
+            return getObject() == null;
+          }
+
+          @Nullable
+          @Override
+          public Object getObject()
+          {
+            if (!unnestListForCurrentRow.isEmpty()) {
+              if (allowSet == null || allowSet.isEmpty()) {
+                return unnestListForCurrentRow.get(index);
+              } else if (allowSet.contains((String) unnestListForCurrentRow.get(index))) {
+                return unnestListForCurrentRow.get(index);
+              }
+            }
+            return null;
+          }
+
+          @Override
+          public Class classOfObject()
+          {
+            return Object.class;
+          }
+        };
+      }
+
+      @Nullable
+      @Override
+      public ColumnCapabilities getColumnCapabilities(String column)
+      {
+        if (!outputName.equals(columnName)) {
+          baseColumSelectorFactory.getColumnCapabilities(column);
+        }
+        return baseColumSelectorFactory.getColumnCapabilities(columnName);
+      }
+    };
+  }
+
+  @Override
+  public DateTime getTime()
+  {
+    return baseCursor.getTime();
+  }
+
+  @Override
+  public void advance()
+  {
+    advanceUninterruptibly();
+    BaseQuery.checkInterrupted();
+  }
+
+  @Override
+  public void advanceUninterruptibly()
+  {
+    do {
+      advanceAndUpdate();
+    } while (matchAndProceed());
+  }
+
+  @Override
+  public boolean isDone()
+  {
+    if (needInitialization && baseCursor.isDone() == false) {
+      initialize();
+    }
+    return baseCursor.isDone();
+  }
+
+  @Override
+  public boolean isDoneOrInterrupted()
+  {
+    if (needInitialization && baseCursor.isDoneOrInterrupted() == false) {
+      initialize();
+    }
+    return baseCursor.isDoneOrInterrupted();
+  }
+
+  @Override
+  public void reset()
+  {
+    index = 0;
+    needInitialization = true;
+    baseCursor.reset();
+  }
+
+  @Override
+  public void initialize()
+  {
+    if (columnValueSelector != null) {
+      this.currentVal = this.columnValueSelector.getObject();
+      this.unnestListForCurrentRow = new ArrayList<>();
+      if (currentVal == null) {
+        unnestListForCurrentRow = new ArrayList<>();
+        unnestListForCurrentRow.add(null);
+      } else {
+        if (currentVal instanceof List) {
+          unnestListForCurrentRow = (List<Object>) currentVal;
+        } else if (currentVal.getClass().equals(String.class)) {
+          unnestListForCurrentRow.add(currentVal);
```

Comment:
Adding more and more things to the list is going to effectively mean that you are keeping a reference to all of the values in memory as you go through the processing.  This is effectively a memory leak and I'm not sure it's necessary?

While it might seem ugly, I think you would be better off with a reference for a single Object and a reference for a List<> and then actually having logic that decides which one it is "iterating" over.